### PR TITLE
Introduce DateTime type

### PR DESCRIPTION
**Pull request by SimonJF**
# Overview
The current support for dates and times in Links is a record which contains the current date and time as a record. This is difficult to compare, doesn't work well with time zones, and doesn't work with native database types.

This patch introduces a new primitive type, `DateTime`, which represents timezone-aware dates and times, and is a prerequisite for merging the temporal database work. The patch is substantial since it needs to work across all three tiers. 

# Usage

* Obtain a DateTime via...
  * the `now()` function to get a timestamp for the current local time 
  * Using `parseDate` on an ISO-formatted string (e.g., `parseDate("26-07-2021 14:26:00+1")`)
  * A `DateTime` field in the database
  * `intToDate(X)` where `X` is a UNIX timestamp
  * `beginning_of_time` and `forever`, which are special timestamps guaranteed to be less than (resp. greater than) all other timestamps

* Project fields out of the type:
  * `utcYear`, `utcMonth`, `utcDay`, `utcHours`, `utcMinutes`, `utcSeconds` projects the given field in the UTC time zone
  * `localYear`, `localMonth`, `localDay`, `localHours`, `localMinutes`, `localSeconds` projects the given field in the local time zone
  * `dateYear`, `dateMonth`, `dateDay`, `dateHours`, `dateMinutes`, `dateSeconds` projects the given field in a given timezone (e.g., to project the hours field of a `DateTime` `dt` in BST, one would write `dateHours(dt, 1)`, where `1` is the timezone offset.

Other stuff:
  * Print the `DateTime` using `show`, `showLocal`, and `showUTC`
  * Compare the `DateTime`s as normal

# Design

Timezones make things entertaining:

  * Parsing a timestamp without a timezone offset on the client or the server assumes the timestamp is in local time
  * Parsing a timestamp without a timezone offset on the database assumes the timestamp is in UTC
  * Internally the timestamp is stored as UTC, and always transferred between client and server in UTC

`DateTime` is deliberately a primitive and not an abstract type. This makes it far easier to do the database integration.

# Limitations
  * This patch doesn't support lenses yet.
  * Since only PostgreSQL supports `infinity` and `-infinity`, we only support PostgreSQL at present.
  * Because of limitations in the underlying libraries, the maximum timezone granularity is an hour. Therefore, we do not support , e.g., Indian timezones

# Testing / Examples
  * Server unit tests: `tests/datetime.tests`
  * Client application showcasing client functions and client<->server serialisation `examples/mvu/dates.links`
  * Database tests: `tests/shredding/dates.{sql, links}`

### CODE REVIEW

File: core/lib.ml
Comment (dhil) [inline core/lib.ml:1203]:
```suggestion
  "beginningOfTime",
```

Comment (SimonJF) [reply core/lib.ml:1203]: Of course, how silly -- will do

**Comment by SimonJF** [comment]
(Documenting):

I believe a request on the call was:

  * Change the interface for `localSeconds` and `utcSeconds` from `(DateTime) -> Float` to `(DateTime) -> Int`
  * Add `localMilliseconds` and `utcMilliseconds` with signature `(DateTime) -> Int`

This separates the implementation of the OCaml library from the library, and also gives us access to the milliseconds on the client (which as it stands are inaccessible).

There was also a request about `dateToInt`. Rest assured, this is already implemented and accessible:
```
links> dateToInt;;
dateToInt : (DateTime) ~> Int
```

**Comment by jamescheney** [comment]
This obviously will have interactions with #936 (hopefully minor), which has been hanging around a while now and I'd like to get merged without unnecessary delay.

**Comment by SimonJF** [comment]
Absolutely, feel free to get #936 in first; the actual DB stuff here is quite minor.

**Comment by dhil** [comment]
@SimonJF is this patch ready for review?

**Comment by SimonJF** [comment]
Yes, reviews would be greatly appreciated :)

File: core/commonTypes.ml
Comment (dhil) [inline core/commonTypes.ml:61]: I suggest you add `to_string`. I can be an alias of `show` if you like.

File: core/json.ml
Comment (dhil) [inline core/json.ml:136]:
```suggestion
        int_of_float (UnixTimestamp.of_calendar ts) in
```

File: core/lib.ml
Comment (dhil) [inline core/lib.ml:1116]: Changelog: Breaking change.

File: core/lib.ml
Comment (dhil) [inline core/lib.ml:1126]: Changelog: Breaking change.

File: core/lib.ml
Comment (dhil) [inline core/lib.ml:1131]:
```suggestion
  "dateMonth", project_datetime_conv (CalendarShow.month ->- CalendarLib.Date.int_of_month);
```

File: core/lib.ml
Comment (dhil) [inline core/lib.ml:1129]:
```suggestion
  "dateYear", (project_datetime_conv CalendarShow.year);
```

File: core/lib.ml
Comment (dhil) [inline core/lib.ml:1130]:
```suggestion
  "dateDay", project_datetime_conv CalendarShow.day_of_month;
```

File: core/lib.ml
Comment (dhil) [inline core/lib.ml:1132]:
```suggestion
  "dateHours", project_datetime_conv CalendarShow.hour;
```

File: core/lib.ml
Comment (dhil) [inline core/lib.ml:1133]:
```suggestion
  "dateMinutes", project_datetime CalendarShow.minute;
```

File: lib/js/jslib.js
Comment (dhil) [inline lib/js/jslib.js:4450]:
```suggestion
const beginningOfTime = { _type: "-infinity" };
```

File: lib/js/jslib.js
Comment (dhil) [inline lib/js/jslib.js:1170]:
```suggestion
    } else if (TYPES.isDateTime(value)) {
```

File: prelude.links
Comment (dhil) [inline prelude.links:1414]:
```suggestion
      println("Expected:");
```

File: prelude.links
Comment (dhil) [inline prelude.links:1415]:
```suggestion
      println(show(expected));
```

File: prelude.links
Comment (dhil) [inline prelude.links:1416]:
```suggestion
      println("but got:");
```

File: lib/js/jslib.js
Comment (dhil) [inline lib/js/jslib.js:320]:
```suggestion
  isDateTime: function (val) {
```

**Review by dhil — CHANGES_REQUESTED** [review]
LGTM. My change requests are mostly concerned with code base consistency.

File: core/lib.ml
Comment (SimonJF) [reply core/lib.ml:1131]: Much nicer!

**Comment by SimonJF** [comment]
Thanks for the reviews, @dhil. I have acted on your suggestions, and also refined the interface so we can get the milliseconds as a separate field (similar to the design of the JS Date library). I think this is good to go if you're happy, after #936 is merged.

**Review by dhil — APPROVED** [review]
LGTM. I will merge this now as I am about to merge in a string other PRs. I seem to recall to have promised you to merge this PR before other changes.

**Comment by jamescheney** [comment]
I think this was just waiting on #936 which has not merged yet since @wricciot has been away.  The only remaining things to do on that were minor nitpicks I had, so if you can give me an hour or two I can probably just do them and that can be merged first.

**Pull request merged**
